### ISSUE TITLE
Memory leaks in 1.1.3-RC7, 1.2.0

### ISSUE THREAD

**Issue by malirod** [log4cplus/log4cplus#203]
I've crated some basic app to test base functionality: logging with different levels via async file appender.
This app was tested with memory&address sanitizers and valgrind. All tools reported about leaks. Reports attached.
Seems this is some kind of degradation. I've been using some time ago 1.1.3-RC2 and there was clean output from sanitizers and valgrind.
Reports for 1.1.3-RC7:
[asan-warnings.txt](https://github.com/log4cplus/log4cplus/files/541079/asan-warnings.txt)
[msan-warnings.txt](https://github.com/log4cplus/log4cplus/files/541080/msan-warnings.txt)
[valgrind-report.txt](https://github.com/log4cplus/log4cplus/files/541081/valgrind-report.txt)

**Comment by wilx** [comment]
Can you try with the HEAD of the 1.1.x branch? I have recently committed changes to fix one leak: https://github.com/log4cplus/log4cplus/issues/195

**Comment by malirod** [comment]
Sure. Is there any tricks to build this branch right after clone or it just works?

**Comment by malirod** [comment]
I've rested with 1.2.0 and here are reports
[msan-warnings-1.2.0.txt](https://github.com/log4cplus/log4cplus/files/541168/msan-warnings-1.2.0.txt)
[asan-warnings-1.2.0.txt](https://github.com/log4cplus/log4cplus/files/541170/asan-warnings-1.2.0.txt)
[valgring-report-1.2.0.txt](https://github.com/log4cplus/log4cplus/files/541171/valgring-report-1.2.0.txt)

going to retest with HEAD of 1.1.x

**Comment by malirod** [comment]
Tested HEAD of 1.1.x - still leaks
[asan-warnings-1.1.x-HEAD.txt](https://github.com/log4cplus/log4cplus/files/541199/asan-warnings-1.1.x-HEAD.txt)
[msan-warnings-1.1.x-HEAD.txt](https://github.com/log4cplus/log4cplus/files/541200/msan-warnings-1.1.x-HEAD.txt)
[valgrind-report-1.1.x-HEAD.txt](https://github.com/log4cplus/log4cplus/files/541201/valgrind-report-1.1.x-HEAD.txt)

**Comment by wilx** [comment]
@malirod How does your test code look like? Are you calling `log4cplus::Logger::shutdown();` at the end of your test code?

**Comment by malirod** [comment]
@wilx, yes, I'm calling shutdown actually twice. is there any better method to force flush asyc appender when required? I haven't found any better solution but to shutdown. After shutdown I'm not writing to log. I'm going to finish publishing my project on github soon. I'll share full sources.

**Comment by malirod** [comment]
@wilx , here is my project https://github.com/malirod/logger-facade
In readme there is the special section with description how to build and run with address sanitizer

**Comment by malirod** [comment]
@wilx, I've double checked my app (with 1.2.0) with only rotating file logger with no async queue. no leaks this way. Only msan still reports about usage of un-init var. Plus valgrind reports about `still reachable: 72,704 bytes in 1 blocks`, but as far as I understand this is not an real issue. So seems problem in async appender.

**Comment by malirod** [comment]
Seems this issues relates to #205 
Leaks are constant and not depend on actual amount of logged data. So problem in some initialization.
I checked in 1.2.0 some warnings from asan and found that they are related to auto_ptr(there were deprecated because usage can lead to problems). In master there were replaced with unique_ptr so there cannot be leaks, but seems logic is not ready for proper memory management. I think there must be some simple fix for both #205 and #203 (at least on master)

**Comment by wilx** [comment]
So I have run your suite against just released 1.2.1-RC1 and I got this:

```
> ASAN_OPTIONS="detect_leaks=1" ./build/debug/testrunner
[==========] Running 7 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 7 tests from TestLogger
[ RUN      ] TestLogger.LogFromFreeFunctionFromAnonymousNamespace
[       OK ] TestLogger.LogFromFreeFunctionFromAnonymousNamespace (2 ms)
[ RUN      ] TestLogger.LogFromClassMethod
[       OK ] TestLogger.LogFromClassMethod (1 ms)
[ RUN      ] TestLogger.AutoTrace
[       OK ] TestLogger.AutoTrace (1 ms)
[ RUN      ] TestLogger.ChainedOutput
[       OK ] TestLogger.ChainedOutput (1 ms)
[ RUN      ] TestLogger.UseLocalLogger
[       OK ] TestLogger.UseLocalLogger (1 ms)
[ RUN      ] TestLogger.WriteToLogInLoop
[       OK ] TestLogger.WriteToLogInLoop (236 ms)
[ RUN      ] TestLogger.InitFromFileNameAndFlushWhenOutOfScope
[       OK ] TestLogger.InitFromFileNameAndFlushWhenOutOfScope (1 ms)
[----------] 7 tests from TestLogger (243 ms total)

[----------] Global test environment tear-down
[==========] 7 tests from 1 test case ran. (244 ms total)
[  PASSED  ] 7 tests.

```

This is with GCC version 6.2.0 20161005 (Ubuntu 6.2.0-5ubuntu12)

**Comment by wilx** [comment]
And I have just pushed and merged 04d39f48741205e081f6a5a6d95f2d0e10814f73 to fix the same issue on 1.1.x branch.

**Comment by wilx** [comment]
I think the issue is now fixed. Thank you for your test case.
